test/datahike/test/tuples_test.cljc

```diff
@@ -0,0 +1,423 @@
+
+    (d/transact! conn [[:db/add 100 :b "b"]
+                       [:db/add 200 :b "b"]
+                       {:db/id 300 :a "a" :b "B"}])
+
+    (is (= #{[100 :a "a"]
+             [100 :b "b"]
+             [100 :a+b ["a" "b"]]
+             [200 :a "A"]
+             [200 :b "b"]
+             [200 :a+b ["A" "b"]]
+             [300 :a "a"]
+             [300 :b "B"]
+             [300 :a+b ["a" "B"]]}
+          (some-datoms (d/db conn) [100 200 300])))
+
+    (is (thrown-with-msg? ExceptionInfo #"Cannot add .* because of unique constraint: .*"
+          (d/transact! conn [[:db/add 100 :a "A"]])))
+    (is (thrown-with-msg? ExceptionInfo #"Cannot add .* because of unique constraint: .*"
+          (d/transact! conn [[:db/add 100 :b "B"]])))
+    (is (thrown-with-msg? ExceptionInfo #"Cannot add .* because of unique constraint: .*"
+          (d/transact! conn [[:db/add 100 :a "A"]
+                             [:db/add 100 :b "B"]])))
+
+    (testing "multiple tuple updates"
+      ;; changing both tuple components in a single operation
+      (d/transact! conn [{:db/id 100 :a "A" :b "B"}])
+      (is (= {:db/id 100 :a "A" :b "B" :a+b ["A" "B"]}
+            (d/pull (d/db conn) '[*] 100)))
+
+      ;; adding entity with two tuple components in a single operation
+      (d/transact! conn [{:db/id 4 :a "a" :b "b"}])
+      (is (= {:db/id 4 :a "a" :b "b" :a+b ["a" "b"]}
+            (d/pull (d/db conn) '[*] 4))))))
+
+(deftest test-validation
+  (let [db (d/empty-db {:a+b {:db/valueType :db.type/tuple
+                           :db/tupleAttrs [:a :b]}})
+        db1 (d/db-with db [[:db/add 100 :a "a"]])
+        err-msg #"Can’t modify tuple attrs directly:.*"]
+    (is (thrown-with-msg? ExceptionInfo err-msg
+          (d/db-with db [[:db/add 100 :a+b [nil nil]]])))
+    (is (thrown-with-msg? ExceptionInfo err-msg
+          (d/db-with db1 [[:db/add 100 :a+b ["a" nil]]])))
+    (is (thrown-with-msg? ExceptionInfo err-msg
+          (d/db-with db [[:db/add 100 :a "a"]
+                         [:db/add 100 :a+b ["a" nil]]])))
+    (is (thrown-with-msg? ExceptionInfo err-msg
+          (d/db-with db1 [[:db/retract 100 :a+b ["a" nil]]])))))
+
+(deftest test-indexes
+  (let [db (-> (d/empty-db {:a+b+c {:db/tupleAttrs [:a :b :c]
+                                    :db/valueType :db.type/tuple
+                                    :db/index true}})
+             (d/db-with
+               [{:db/id 1 :a "a" :b "b" :c "c"}
+                {:db/id 2 :a "A" :b "b" :c "c"}
+                {:db/id 3 :a "a" :b "B" :c "c"}
+                {:db/id 4 :a "A" :b "B" :c "c"}
+                {:db/id 5 :a "a" :b "b" :c "C"}
+                {:db/id 6 :a "A" :b "b" :c "C"}
+                {:db/id 7 :a "a" :b "B" :c "C"}
+                {:db/id 8 :a "A" :b "B" :c "C"}]))]
+    (is (= [6]
+          (mapv :e (d/datoms db :avet :a+b+c ["A" "b" "C"]))))
+    (is (= []
+          (mapv :e (d/datoms db :avet :a+b+c ["A" "b" nil]))))
+    (is (= [8 4 6 2]
```

Comment:
Is it correct that 4 and 2 are returned here as well? what are the matching rules here? I would have expected to only get back [8 6]

Reply:
It is correct because "A" is smaller than "a" in Java/Clojure (e.g.: `(.compareTo "B" "b") => -32`).
So `["A" "b" "C"]` is smaller than `["A" "b" "c"]`.